psi4/src/psi4/libsapt_solver/exch-ind30.cc

```diff
@@ -314,5 +319,305 @@ double SAPT2p3::exch_ind30_3(double **sBS) {
 
     return (energy);
 }
+
+// Compute the nonapproximated third-order exchange-induction //
+// Konrad Patkowski, based on Jonathan Waldrop's Psi4NumPy code //
+// September 2021 //
+void SAPT2p3::sinf_e30ind() {
+
+    if (print_) {
+        outfile->Printf("  ==> Nonapproximated third-order induction <==\n\n");
+    }
+
+    // => Sizing <= //
+
+    int nn = basisset_->nbf();
+    int na = noccA_;
+    int nb = noccB_;
+    int nr = nvirA_;
+    int ns = nvirB_;
+    int nQ = ribasis_->nbf();
+
+    int nT = 1;
+#ifdef _OPENMP
+    nT = Process::environment.get_n_threads();
+#endif
+
+    double **uAR = block_matrix(na, nr);
+    double **uBS = block_matrix(nb, ns);
+    psio_->read_entry(PSIF_SAPT_AMPS, "Ind30 uAR Amplitudes", (char *)uAR[0], sizeof(double) * na * nr);
+    psio_->read_entry(PSIF_SAPT_AMPS, "Ind30 uBS Amplitudes", (char *)uBS[0], sizeof(double) * nb * ns);
+
+    // => Intermolecular overlap matrix and inverse <= //
+    std::shared_ptr<Matrix> Sab = linalg::triplet(CoccA_, Smat_, CoccB_, true, false, false);
+
+    double** Sabp = Sab->pointer();
+    auto D = std::make_shared<Matrix>("D", na + nb, na + nb);
+    D->identity();
+    double** Dp = D->pointer();
+    for (int a = 0; a < na; a++) {
+        for (int b = 0; b < nb; b++) {
+            Dp[a][b + na] = Dp[b + na][a] = Sabp[a][b];
+        }
+    }
+    D->power(-1.0, 1.0E-12);
+    Dp = D->pointer();
+
+    // => New Stuff <= //
+    // Start with T's
+    std::shared_ptr<Matrix> Sbr = linalg::triplet(CoccB_, Smat_, CvirA_, true, false, false);
+    std::shared_ptr<Matrix> Sas = linalg::triplet(CoccA_, Smat_, CvirB_, true, false, false);
+    auto Tar = std::make_shared<Matrix>("Tar", na, nr);
+    auto Tbr = std::make_shared<Matrix>("Tbr", nb, nr);
+    auto Tas = std::make_shared<Matrix>("Tas", na, ns);
+    auto Tbs = std::make_shared<Matrix>("Tbs", nb, ns);
+
+    C_DGEMM('N', 'N', na, nr, nb, 1.0, &Dp[0][na], na + nb, Sbr->pointer()[0], nr, 0.0,
```

Comment:
The `libmints` function gets block allows you to get a block of a Matrix. You _could_ use this to get the four blocks of the D^-1 matrix as their own matrices, and then 8 of the "New stuff" DGEMM calls become simple doublets. I expect your half-transforms and your xi-dependent terms simplify as well. There will be some additional copy overhead from this approach, but the code becomes cleaner. I won't insist on you doing this, but I do recommend you consider it.

Reply:
I see your point, it's just that I'm quite confused about the syntax of the slice definition for `matrix->get_block`.